packages/smart/contracts/turbo/AMMFactory.sol

```diff
@@ -125,11 +126,34 @@ contract AMMFactory {
         uint256 _sets = _marketFactory.calcShares(_collateralIn);
         _marketFactory.mintShares(_marketId, _sets, address(this));
 
-        // Add liquidity to pool
-        uint256 _totalLPTokens = 0;
+        // Find poolAmountOut
+        uint256 _poolAmountOut = MAX_UINT;
+
+        {
+            uint256 _totalSupply = _pool.totalSupply();
+            uint256[] memory _maxAmountsIn = new uint256[](_market.shareTokens.length);
+            for (uint256 i = 0; i < _market.shareTokens.length; i++) {
+                _maxAmountsIn[i] = _sets;
+
+                OwnedERC20 _token = _market.shareTokens[i];
+                uint256 _bPoolTokenBalance = _pool.getBalance(address(_token));
+                uint _tokenPoolAmountOut = bdiv(bmul(_sets, _totalSupply), _bPoolTokenBalance);
+
+                if(_tokenPoolAmountOut < _poolAmountOut) {
+                    _poolAmountOut = _tokenPoolAmountOut;
+                }
+            }
+            _pool.joinPool(_poolAmountOut, _maxAmountsIn);
+        }
+
+        // Add liquidity to pool by depositing the remaining share tokens.
+        uint256 _totalLPTokens = _poolAmountOut;
         for (uint256 i = 0; i < _market.shareTokens.length; i++) {
             OwnedERC20 _token = _market.shareTokens[i];
-            uint256 __acquiredLPTokens = _pool.joinswapExternAmountIn(address(_token), _sets, 0);
+            uint256 _tokenBalance = _token.balanceOf(address(this));
+            if(_tokenBalance == 0) continue;
+
+            uint256 __acquiredLPTokens = _pool.joinswapExternAmountIn(address(_token), _token.balanceOf(address(this)), 0);
```

Comment:
I worry that this will still fail under certain conditions: the balances differ from the weights and the liquidity is low. Especially if one of the balances is very low.

Reply:
We talked about it. There will need to be a dynamic minimum liquidity the user can add, depending on these factors. The calculation isn't too difficult. The workaround is for users themselves to mint sets and to call joinPool on the bpool.

The alternative is to make the typical UX flow include two methods of adding liquidity. That might be the right path... but let's try this one first.

Reply:
Will resolve this in another pr. We can check for liquidity on the bpool and if it is too low just transfer the shares back to `msg.sender`.